perf(core): avoid storing LView in __ngContext__

crisbeto opened the pull request:
Currently we save a reference to an `LView` on most DOM nodes created by Angular either by saving the `LView` directly in the `__ngContext__` or by saving the `LContext` which has a reference to the `LView`. This can be a problem if the DOM node is retained in memory, because the `LView` has references to all of the child nodes of the view, as well as other internal data structures.

Previously we tried to resolve the issue by clearing the `__ngContext__` when a node is removed (see https://github.com/angular/angular/pull/36011), but we decided not to proceeed, because it can slow down destruction due to a megamorphic write.

These changes aim to address the issue while reducing the performance impact by assigning a unique ID when an `LView` is created and adding it to `__ngContext__`. All active views are tracked in a map where their unique ID is used as the key. We don't need to worry about leaks within that map,  because `LView`s are an internal data structure and we have complete control over when they are  created and destroyed.

Fixes #41047.

crisbeto commented on packages/core/src/render3/instructions/lview_tracking.ts:
```diff
@@ -0,0 +1,32 @@
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
```
I had to define these functions in a separate file to avoid circular imports.

crisbeto commented on packages/core/src/render3/instructions/lview_tracking.ts:
```diff
@@ -0,0 +1,32 @@
+// LViews that are currently being tracked. Each index in the array
+// corresponds to the ID of an LView. When an LView is no longer being
+// tracked, its index is replaced with null so that it can't be reused.
+const TRACKED_LVIEWS: (LView|null)[] = [];
```
@mhevery we talked about using a `Map` for tracking these, but I ended up going for the array, because we don't use `Map` anywhere else in the Ivy runtime and it may require people to add a polyfill.

AndrewKushnir replied:
If I understand correctly, the `Map` is supported in IE 11 (and all modern browsers):

https://caniuse.com/?search=Map()

I'd vote for using `Map` here, so that there are no memory implications of having ever-growing array.

gkalpak replied:
Beware that IE11 only has partial support for `Map`s out-of-the-box. But we do list ES2015 polyfills in the list of [mandatory polyfills][1], so it might be OK to assume `Map` support (either native or via a polyfill).

[1]: https://angular.io/guide/browser-support#mandatory-polyfills

crisbeto replied:
Ah nice, I was under the impression that IE doesn't support it. I've switched it to a `Map`, although I wonder whether that won't end up being the same under the hood. One of the simpler ways to implement a `Map` from scratch is to use two arrays internally (keys and values), but I'm guessing that the browser is doing something smarter.

mhevery replied:
Yes, please let's use Map here, otherwise the array will grow forever.

crisbeto commented on packages/core/test/bundling/forms_reactive/index.ts:
```diff
@@ -93,5 +93,15 @@ class FormsExampleModule {
+// against it will load a different core bundle. These symbols are exposed so that they
+// can interact with the correct `@angular/core` instance.
+module.exports = {
+  whenRendered,
```
These tests revealed a "fun" bug: the test that are run against these bundles use a different `@angular/core` compared to the the bundle itself which means that the runtime state isn't shared. We just hadn't noticed it until now, because the entire `LView` was inside `__ngContext__`.

mhevery replied:
Nice! thanks for the fix

mhevery commented on packages/core/src/render3/interfaces/context.ts:
```diff
@@ -23,9 +22,9 @@ import {LView} from './view';
+   * ID of the component's parent view data.
    */
-  lView: LView;
+  lViewId: number;
```
I think it would be cleaner if we left this interface as is, and have the `getLContext` method do the lookup from id to `Context`. I think doing that change would make it cleaner as fewer places would have to change.

crisbeto replied:
Storing the `LView` here would partially defeat the purpose of these changes, because the entire `LContext` can be stored on `__ngContext__`.

mhevery replied:
But we could implement `lView` as a getter so that the interface stays the same but the relationship is broken.

mhevery commented on packages/core/src/render3/context_discovery.ts:
```diff
@@ -181,7 +186,8 @@ const MONKEY_PATCH_KEY_NAME = '__ngContext__';
   ngDevMode && assertDefined(target, 'Target expected');
-  target[MONKEY_PATCH_KEY_NAME] = data;
+  // Only attach the ID of the view in order to avoid memory leaks (see #41047).
+  target[MONKEY_PATCH_KEY_NAME] = Array.isArray(data) ? data[ID] : data;
```
Can we use `isLView` instead of `Array.isArray`

crisbeto replied:
Changed. I did it because we had a few other places in the same files that were asserting it in the same way.

mhevery commented on packages/core/test/acceptance/i18n_spec.ts:
```diff
@@ -652,12 +652,12 @@ onlyInIvy('Ivy i18n logic').describe('runtime i18n', () => {
         fixture.detectChanges();
         expect((fixture.nativeElement as Element).textContent).toEqual('just now');
-        expect(lViewDebug.nodes.map(toTypeContent)).toEqual(['IcuContainer(<!--ICU 20:0-->)']);
+        expect(lViewDebug.nodes.map(toTypeContent)).toEqual(['IcuContainer(<!--ICU 21:0-->)']);
```
Hmm... What did the ICU ids change?

crisbeto replied:
It's probably because they depend on the `HEADER_OFFSET` of the `LView` which was increased by 1.

crisbeto commented:
All of the feedback should be addressed.

AndrewKushnir commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -325,9 +337,9 @@ export interface Listener {
 export function getListeners(element: Element): Listener[] {
   assertDomElement(element);
```
Pre-existing assert without `ngDevMode`.
@crisbeto it's outside of the scope of your PR, but it you could help to add `ngDevMode` here as well, that'd be great!

AndrewKushnir commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -325,9 +337,9 @@ export interface Listener {
   assertDomElement(element);
   const lContext = loadLContext(element, false);
-  if (lContext === null) return [];
+  const lView = lContext ? getLViewById(lContext.lViewId) : null;
```
```suggestion
  const lView = lContext !== null ? getLViewById(lContext.lViewId) : null;
```

AndrewKushnir reviewed (commented):
Thanks for the changes @crisbeto 👍 Just left a few comments.

One more nit: should we change the `fix(core)` -> `refactor(core)` or `perf(core)`? It looks like this is more of an improvement vs a bug fix.

mgechev commented on packages/core/test/acceptance/debug_spec.ts:
```diff
@@ -9,6 +9,7 @@
 import {Component} from '@angular/core';
 import {getLContext} from '@angular/core/src/render3/context_discovery';
 import {LViewDebug} from '@angular/core/src/render3/instructions/lview_debug';
+import {getLViewById} from '@angular/core/src/render3/instructions/lview_tracking';
```
DevTools currently relies on `__ngContext__` to get metadata. Ideally, we'd want to abstract this with debugging APIs within `LViewDebug`, but we haven't finalized the timeline for this.

I'd suggest exporting `getLViewById` as a private so that DevTools has a fallback to access the `LView`.

crisbeto replied:
I don't mind exporting it, but I want to understand the usage first. Is it specifically using the `LView` or the `LContext` from `__ngContext__`? Maybe the same can be achieved with one of the APIs in `window.ng`. 

I'm not sure that `LView` is the most stable API in the first place, because the `HEADER_OFFSET` can change without warning when we add a new field.

mgechev replied:
Currently, we access the `__ngContext__` of the root element and from there we extract the logical tree. We were originally using the `window.ng` APIs but they are too slow for the purpose of DevTools.

There's a work in progress [PR](https://github.com/angular/angular/pull/41144), which aims to move the logical/render tree construction outside of DevTools so we don't rely on `HEADER_OFFSET`, but we'd still need access to `LViewDebug`.

Until we're able to implement the extension of `LViewDebug` we'd have to rely on the plain `LView`. To mitigate the risk of different constants changing, we've been testing with the latest build artifacts so we can catch regressions and post a patch before we break people.

mgechev reviewed (changes requested):
The change will break DevTools. Exporting `getLViewById` will be helpful to let our metadata collection mechanism fallback to the new approach.

crisbeto commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -76,7 +79,8 @@ export function getComponent<T>(element: Element): T|null {
 export function getContext<T>(element: Element): T|null {
   assertDomElement(element);
```
Added for consistency, but it doesn't really matter since these functions only exist in dev mode.

crisbeto committed:
perf(core): avoid storing LView in __ngContext__

Currently we save a reference to an `LView` on most DOM nodes created by Angular either by saving
the `LView` directly in the `__ngContext__` or by saving the `LContext` which has a reference to
the `LView`. This can be a problem if the DOM node is retained in memory, because the `LView` has
references to all of the child nodes of the view, as well as other internal data structures.

Previously we tried to resolve the issue by clearing the `__ngContext__` when a node is removed
(see https://github.com/angular/angular/pull/36011), but we decided not to proceeed, because it can
slow down destruction due to a megamorphic write.

These changes aim to address the issue while reducing the performance impact by assigning a unique
ID when an `LView` is created and adding it to `__ngContext__`. All active views are tracked in
a map where their unique ID is used as the key. We don't need to worry about leaks within that map,
because `LView`s are an internal data structure and we have complete control over when they are
created and destroyed.

Fixes #41047.

crisbeto committed:
fixup! perf(core): avoid storing LView in __ngContext__

AndrewKushnir reviewed (approved):
LGTM, thanks for addressing the feedback @crisbeto 👍 

FYI, I'm adding the "blocked" label for now to make sure we unblock Minko ([based on his comment](https://github.com/angular/angular/pull/41358/files#r603686402)) before merging.

crisbeto committed:
fixup! perf(core): avoid storing LView in __ngContext__

crisbeto commented:
The latest set of feedback has been addressed.

AndrewKushnir commented:
Hi @crisbeto, quick update: we had a chat with @mgechev and it looks like this PR is now unblocked and we can land it in RC (**not** in the patch branch though) and it would not affect DevTools. DevTools extension has a logic that does the following:

- if `ng.getDirectiveMetadata` method is present (introduced in `v12.0.0`), DevTools would use that and would not rely on __ngContext__ at all (which is great)
- otherwise (for v11.2.x versions and older), there is no `ng.getDirectiveMetadata` method available, so the logic falls back to using __ngContext__

So landing this change in RC branch (and master branch) should be ok from DevTools perspective.

Thank you.

crisbeto committed:
fixup! perf(core): avoid storing LView in __ngContext__

crisbeto commented:
I've rebased, resolved the CI issues and removed the public export of `getLViewById`.

Also cc @mgechev and @mhevery for approvals.

crisbeto commented:
Changed `lView` to a getter as discussed.

mhevery commented on packages/core/src/render3/interfaces/context.ts:
```diff
@@ -21,35 +22,41 @@ import {LView} from './view';
+export class LContext {
+  /** Component's parent view data. */
+  get lView(): LView|null {
+    return getLViewById(this._lViewId);
```
I think we should change this to throw if `null` What do you think?

Then you can revert all of the `.lvView!` and `assert` statements

crisbeto replied:
As discussed, I'll leave it as is, because we have at least one place where `null` is an allowed value. See https://github.com/angular/angular/pull/41358/commits/302c7d031ce5f0eaef9c8c9518b54c493eae09ab#diff-92e488b29fcb339dca89bcc326131e4e7482ceac84387f5e5250c3441a9dd326R1912.

crisbeto committed:
fixup! perf(core): avoid storing LView in __ngContext__

AndrewKushnir commented on packages/core/src/render3/instructions/lview_tracking.ts:
```diff
@@ -0,0 +1,35 @@
+import {ID, LView} from '../interfaces/view';
+
+// Keeps track of the currently-active LViews.
+const TRACKED_LVIEWS = new Map<number, LView>();
```
If I understand correctly, having this Map (and the `uniqueIdCounter` as well) scoped to a particular Angular core instance (vs making the Map global) would not allow using functions from the `ng` namespace in case there are multiple Angular apps (with packaged `@angular/core` code) on a page, where each instance exposes debug utilities in `ng` namespace.

crisbeto replied:
Yes, this is what we discussed previously. I decided to hold off on it for now, because it's much more of an edge case now that the dev tools won't depend on `getLViewById` and storing the information will take more code to get right. If this becomes a problem, it can be fixed later.

AndrewKushnir reviewed (commented):
@crisbeto thanks for updating the PR 👍 Please see comments below.

jessicajaniuk reviewed (approved):
LGTM 🍪 

reviewed-for: size-tracking

petebacondarwin commented on packages/core/src/render3/interfaces/context.ts:
```diff
@@ -21,35 +22,41 @@ import {LView} from './view';
+      /**
+       * ID of the component's parent view data.
+       */
+      private _lViewId: number,
```
@AndrewKushnir - when you say it is public... it has a private keyword in front of it...?

AndrewKushnir replied:
@petebacondarwin sorry, that was a typo, I meant to say "private" :)

crisbeto commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -1906,9 +1908,12 @@ export function scheduleTick(rootContext: RootContext, flags: RootContextFlags)
-    const lView = readPatchedLView(rootComponent)!;
-    const tView = lView[TVIEW];
-    renderComponentOrTemplate(tView, lView, tView.template, rootComponent);
+    const lView = readPatchedLView(rootComponent);
+    // We might not have an `LView` if the component was destroyed.
+    if (lView !== null) {
+      const tView = lView[TVIEW];
+      renderComponentOrTemplate(tView, lView, tView.template, rootComponent);
+    }
```
I've added one now, but it's a little contrived.

crisbeto commented on packages/core/src/render3/interfaces/context.ts:
```diff
@@ -21,35 +22,41 @@ import {LView} from './view';
+       * The map of local references (local reference name => element or directive instance) that
+       * exist on this element.
+       */
+      public localRefs?: {[key: string]: any}|null) {}
```
They aren't passed to the constructor, but they can be assigned later on. I'll remove them from the constructor.

crisbeto committed:
fixup! perf(core): avoid storing LView in __ngContext__

crisbeto reviewed (commented):
I've addressed the latest set of feedback.

petebacondarwin reviewed (approved):
Fantastic PR @crisbeto 

Reviewed-for: size-tracking

AndrewKushnir reviewed (approved):
LGTM, thanks for the updates @crisbeto 👍

AndrewKushnir commented:
[Presubmit](http://test/OCL:369724594:BASE:369724629:1619037368246:a640b270).

AndrewKushnir commented:
@crisbeto FYI some newly-added asserts are causing some targets in g3 to fail. I will investigate it further and try to come up with a repro. Adding "blocked" label for now.

AndrewKushnir commented:
@crisbeto quick update: I was able to repro the issue locally, here is my diff:

```
--- a/packages/core/test/acceptance/discover_utils_spec.ts
+++ b/packages/core/test/acceptance/discover_utils_spec.ts
@@ -74,9 +74,16 @@ onlyInIvy('Ivy-specific utilities').describe('discovery utils', () => {
       <child dirA #child></child>
       <child dirA *ngIf="true"></child>
       <ng-container><p></p></ng-container>
+      <ng-container *ngIf="visible">
+        <b>
+          Bold
+          <u>Underline</u>
+        </b>
+      </ng-container>
     `
   })
   class MyApp {
+    visible = true;
     text: string = 'INIT';
     @Input('a') b = 2;
     @Output('c') d = new EventEmitter();
@@ -89,6 +96,14 @@ onlyInIvy('Ivy-specific utilities').describe('discovery utils', () => {
     }
   }
 
+  fit('try to use debug functions on removed elements', () => {
+    const u = fixture.nativeElement.querySelector('u');
+    const uTokens = getInjectionTokens(u);
+    fixture.componentInstance.visible = false;
+    fixture.detectChanges();
+    const uTokens2 = getInjectionTokens(u);  // <-- this will throw
+  });
```

The problem is that the first `getInjectionTokens` call creates a new `LContext` for the `<u>` element (as it walks up the parent tree in `readPatchedData`), which also contains an `lViewId` reference. Once the view gets removed and the necessary cleanup happens (thus an `lView` with a previously stored id can not be found in the map) and we try to invoke the `getInjectionTokens` function again, we try to pull an `lView` using id that no longer has corresponding value in the map. Prior to the change in this PR, there was a direct `LView` reference, so the problem didn't happen. Now we would not be able to establish this connection and retrieve any info using DOM elements removed from the tree. This makes sense to me that the debug info is not available for removed elements, but there might be some code that relies on this (we'd need to run a global presubmit) and also that might be  a change in behavior between Ivy and ViewEngine.

It looks like we need to account for this case (that `lViewId` may point to a removed `lView`) in the code and replace an assert that verifies that the `context.lView` is an instance of an `LView` with a check that would return the default value for a particular function (an empty array or `null`).

Thank you.

crisbeto committed:
fixup! perf(core): avoid storing LView in __ngContext__

crisbeto commented:
Thank you for investigating @AndrewKushnir. I've incorporated your test and similar tests for all the other `window.ng` functions. I've also fixed a few errors similar to the one you found in the other functions.

crisbeto commented:
I've pushed another change to account for the recent switch from `loadLContext` to `getLContext`.

crisbeto committed:
fixup! perf(core): avoid storing LView in __ngContext__

AndrewKushnir commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -122,7 +125,8 @@ export function getOwningComponent<T>(elementOrDir: Element|{}): T|null {
  */
 export function getRootComponents(elementOrDir: Element|{}): {}[] {
-  return [...getRootContext(elementOrDir).components];
+  const lView = readPatchedLView(elementOrDir);
+  return lView ? [...getRootContext(lView).components] : [];
```
Not related to this PR, but it looks like we should use the `getLContext` here too. For non-host elements (which don't have `__ngContext__` patched), this method would return an empty array (vs trying to walk up the tree to find a parent element with the `__ngContext__` field).

crisbeto replied:
Do you mean instead of the `getRootContext` call? `getContext` seems to return a different value.

AndrewKushnir commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -122,7 +125,8 @@ export function getOwningComponent<T>(elementOrDir: Element|{}): T|null {
  */
 export function getRootComponents(elementOrDir: Element|{}): {}[] {
-  return [...getRootContext(elementOrDir).components];
+  const lView = readPatchedLView(elementOrDir);
+  return lView ? [...getRootContext(lView).components] : [];
```
```suggestion
  const lView = readPatchedLView(elementOrDir);
  return lView !== null ? [...getRootContext(lView).components] : [];
```

AndrewKushnir commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -297,7 +302,9 @@ export function getLocalRefs(target: {}): {[key: string]: any} {
 
   if (context.localRefs === undefined) {
-    context.localRefs = discoverLocalRefs(context.lView, context.nodeIndex);
+    const lView = context.lView!;
+    ngDevMode && assertLView(lView);
```
Can the `context.lView` be `null` here (for "removed elements" use-case)?

(similar situations also in `getComponentLView` and `getDebugNode` below)

crisbeto replied:
Changes this and `getDebugNode`, but I kept `getComponentLView` as is since it makes sense to have the assertion there, because the expected return value is `LView`.

AndrewKushnir reviewed (commented):
@crisbeto thanks for applying fixes and rebasing the PR! 👍 
I've left a couple comments (mostly around the areas that might potentially have `null` where we extract `lView` from the context).

crisbeto commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -52,12 +53,14 @@ import {getTNode, unwrapRNode} from './view_utils';
   if (context.component === undefined) {
-    context.component = getComponentAtNodeIndex(context.nodeIndex, context.lView);
+    const lView = context.lView!;
+    ngDevMode && assertLView(lView);
```
It is, but I left it in because it didn't break the newly-added unit tests. It worked, because the `context.component` was being set on a previous call which meant that it never got into the `if` block after the node is destroyed.

crisbeto replied:
It is, but I left it in because it didn't break the newly-added unit tests. It worked, because the `context.component` was being set on a previous call which meant that it never got into the `if` block after the node is destroyed. I'll change it now.

crisbeto committed:
fixup! perf(core): avoid storing LView in __ngContext__

crisbeto reviewed (commented):
I've addressed the latest set of feedback.

AndrewKushnir commented:
[Presubmit](http://test/OCL:369724594:BASE:370117510:1619201517974:ba63549a).

AndrewKushnir commented:
FYI presubmits went well, I'll run a global one during the weekend (to make sure there are no corner-cases that we've not reviewed/discussed) and I will share results on Monday.

AndrewKushnir commented:
[Global Presubmit (Ivy version)](https://test.corp.google.com/ui#id=OCL:369724594:BASE:370383936:1619410227304:5e79f2a1).

AndrewKushnir commented:
FYI, global presubmit went well too, removing the "presubmit" label.
